### PR TITLE
Add separate validation classes for the qr requests

### PR DESCRIPTION
**Pull request by daanstokhof**
multiple requests use this model as reference. each of those request has a different set of required properties.

### CODE REVIEW

File: src/Request/AbstractRequest.php [hunk 1]
```diff
@@ -564,7 +564,7 @@ protected function flattenErrors(array $errors, string $context = ''): array
         foreach ($errors as $key => $value) {
             $return[$key] = $value;
             if (true === is_array($value)) {
-                $return = $this->flattenErrors($value, ltrim($context . ".$key", '.'));
+                $return = $this->flattenErrors($value, ltrim($context . ".{$key}", '.'));
```
Comment (IvoWams) [inline src/Request/AbstractRequest.php:567]: What does this change do ?

File: src/Request/ConfigProvider.php [hunk 2]
```diff
@@ -4,11 +4,11 @@
 
 namespace PayNL\Sdk\Request;
 
-use PayNL\Sdk\{
-    Config\ProviderInterface as ConfigProviderInterface,
+use PayNL\Sdk\{Config\ProviderInterface as ConfigProviderInterface,
```
Comment (IvoWams) [inline src/Request/ConfigProvider.php:7]: Add enter after the {

File: src/Validator/ConfigProvider.php [hunk 3]
```diff
@@ -4,10 +4,10 @@
 
 namespace PayNL\Sdk\Validator;
 
-use PayNL\Sdk\{
-    Common\ManagerFactory,
-    Config\ProviderInterface as ConfigProviderInterface
-};
+use PayNL\Sdk\{Common\ManagerFactory,
```
Comment (IvoWams) [inline src/Validator/ConfigProvider.php:7]: Add enter after the {

File: src/Validator/Manager.php [hunk 4]
```diff
@@ -19,4 +19,19 @@ class Manager extends AbstractPluginManager
...
      */
     protected $instanceOf = ValidatorInterface::class;
+
+    /**
+     * Returns the validator specified in the config if found. if not return the default RequiredMembers validator
+     *
+     * @param $request
+     * @return ValidatorInterface
+     */
+    public function getValidatorByRequest($request) : ValidatorInterface {
```
Comment (IvoWams) [inline src/Validator/Manager.php:29]: Could probably add AbstractRequest as type

File: src/Validator/Qr/Decode.php [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class Decode extends RequiredMembers
+{
+    public function isValid($filledObjectToCheck): bool
+    {
+        $requiredMembers = array('secret', 'uuid');
+        $this->className = get_class($filledObjectToCheck);
+        $data = $this->getDataFromObject($filledObjectToCheck);
+
+        return $this->validate(array_flip($requiredMembers), $data);
```
Comment (IvoWams) [inline src/Validator/Qr/Decode.php:22]: Why the array_flip? Can you not order the array correctly at line 18 ?

File: src/Validator/Qr/Encode.php [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+ */
+class Encode extends RequiredMembers
+{
+    public function isValid($filledObjectToCheck): bool
+    {
+        $requiredMembers = array('secret');
+        $this->className = get_class($filledObjectToCheck);
+        $data = $this->getDataFromObject($filledObjectToCheck);
+
+        return $this->validate(array_flip($requiredMembers), $data);
```
Comment (IvoWams) [inline src/Validator/Qr/Encode.php:22]: why the array_flip ?

File: src/Validator/RequiredMembers.php [hunk 7]
```diff
@@ -37,19 +37,29 @@ class RequiredMembers extends AbstractValidator implements HydratorAwareInterfac
         self::MSG_EMPTY_MEMBERS   => 'Members "%s" within %s are required and therefore cannot be empty',
     ];
 
+    protected $className;
```
Comment (IvoWams) [inline src/Validator/RequiredMembers.php:40]: I'd prefer it if this member was instead just passed as an parameter to validate(...). A classname in itself has nothing to do with the RequiredMembers class.

File: tests/unit/Validator/Qr/DecodeTest.php [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    }
+
+    /**
+     * @return void
+     */
+    public function testItCanValidateWithMissingMembers(): void
+    {
+        $this->setHydrator();
+
+        $data = $this->validator->isValid($this->dummy);
```
Comment (IvoWams) [inline tests/unit/Validator/Qr/DecodeTest.php:104]: This dummy is not missing the members. Make a new dummy with no members.

File: tests/unit/Validator/Qr/EncodeTest.php [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    /**
+     * @return void
+     */
+    public function testItCanValidateWithMissingMembers(): void
+    {
+        $this->setHydrator();
+
+        $data = $this->validator->isValid($this->dummy);
```
Comment (IvoWams) [inline tests/unit/Validator/Qr/EncodeTest.php:103]: Same as before, dummy has no missing members.

File: src/Request/AbstractRequest.php (hunk 1 above)
Comment (daanstokhof) [reply src/Request/AbstractRequest.php:567]: not a lot. it looks nicer imo.

File: src/Validator/Manager.php [hunk 10]
```diff
@@ -19,4 +20,19 @@ class Manager extends AbstractPluginManager
...
      */
     protected $instanceOf = ValidatorInterface::class;
+
+    /**
+     * Returns the validator specified in the config if found. if not return the default RequiredMembers validator
+     *
+     * @param $request
+     * @return ValidatorInterface
+     */
+    public function getValidatorByRequest(AbstractRequest $request) : ValidatorInterface {
```
Comment (isneep) [inline src/Validator/Manager.php:30]: PSR-2!

File: src/Validator/Manager.php [hunk 11]
```diff
@@ -19,4 +20,19 @@ class Manager extends AbstractPluginManager
      * @see AbstractPluginManager::$instanceOf
      */
     protected $instanceOf = ValidatorInterface::class;
+
+    /**
+     * Returns the validator specified in the config if found. if not return the default RequiredMembers validator
+     *
+     * @param $request
```
Comment (isneep) [inline src/Validator/Manager.php:27]: Add type to param doc

File: src/Validator/Manager.php [hunk 12]
```diff
@@ -19,4 +20,19 @@ class Manager extends AbstractPluginManager
...
+     * @return ValidatorInterface
+     */
+    public function getValidatorByRequest(AbstractRequest $request) : ValidatorInterface {
+        $options = $request->getOptions();
+        if(isset($options['validator'])) {
```
Comment (isneep) [inline src/Validator/Manager.php:32]: Due to the fact that the validator can be given through the configuration check in the manager if the class exists:
```php
if (true === isset($options['validator'])) {
    if (true === is_string($options['validator']) && true === $this->has($options['validator'])) {
        return $this->get($options['validator']);
    } elseif (false === is_callable($options['validator']) {
        throw new RuntimeException(....) // TODO: make sure you use the correct exception class
    }
    return ($options['validator'])();
}
return $this->get('RequiredMembers');
```

File: src/Validator/Qr/Decode.php [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+use PayNL\Sdk\Validator\RequiredMembers;
+
+/**
+ * Class RequiredMembers
+ *
+ * @package PayNL\Sdk\Validator
+ */
+class Decode extends RequiredMembers
+{
+    public function isValid($filledObjectToCheck): bool
```
Comment (isneep) [inline src/Validator/Qr/Decode.php:16]: DocBlock

File: src/Validator/Qr/Encode.php [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+use PayNL\Sdk\Validator\RequiredMembers;
+
+/**
+ * Class RequiredMembers
+ *
+ * @package PayNL\Sdk\Validator
+ */
+class Encode extends RequiredMembers
+{
+    public function isValid($filledObjectToCheck): bool
```
Comment (isneep) [inline src/Validator/Qr/Encode.php:16]: DocBlock

File: tests/_support/TestAsset/DummyQr.php [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+use JsonSerializable;
+
+/**
+ * Class DummyQr
+ *
+ * @package Codeception\TestAsset
+ */
+class DummyQr implements DummyInterface, JsonSerializable
+{
+    protected $options = [];
```
Comment (isneep) [inline tests/_support/TestAsset/DummyQr.php:16]: Add DocBlocks for vars and methods, even if it is a test asset

File: tests/unit/Validator/ManagerTest.php [hunk 16]
```diff
@@ -21,12 +24,18 @@ class ManagerTest extends UnitTest
...
     /**
      * @return void
      */
     public function _before(): void
     {
+        /** @var Manager $manager */
         $this->manager = new Manager();
+
+        /** @var Application $application */
+        $this->application = Application::init($this->tester->getConfig());
```
Comment (isneep) [inline tests/unit/Validator/ManagerTest.php:38]: Application is already initialized by the SDK codeception Module, it's not necessary to initiate it again. Use the helper methods to get the application, service manager, etc.

File: tests/unit/Validator/ManagerTest.php [hunk 17]
```diff
@@ -50,4 +59,33 @@ public function testItHasADefinedInstanceOfAttribute(): void
         verify($instanceOf)->notEmpty();
         verify($instanceOf)->equals(ValidatorInterface::class);
     }
+
+    /**
+     * @return void
+     */
+    public function testItCanGetCustomValidatorByRequest(): void
+    {
+        $this->application->setRequest('DecodeQr');
```
Comment (isneep) [inline tests/unit/Validator/ManagerTest.php:68]: Requests to test should be mocked

File: tests/unit/Validator/ManagerTest.php [hunk 18]
```diff
@@ -50,4 +59,33 @@ public function testItHasADefinedInstanceOfAttribute(): void
...
+        $options =  $this->application->getRequest()->getOptions();
+        $options['validator'] = Decode::class;
```
Comment (isneep) [inline tests/unit/Validator/ManagerTest.php:70]: Not necessary, should me modified in the configuration for test(s)

File: tests/unit/Validator/ManagerTest.php [hunk 19]
```diff
@@ -50,4 +59,33 @@ public function testItHasADefinedInstanceOfAttribute(): void
...
+    }
+
+    /**
+     * @return void
+     */
+    public function testItCanGetDefaultValidatorByRequest(): void
+    {
+        $this->application->setRequest('GetCurrency', ['currencyId' => 'EUR'], [], ['Currency' => []]);
+
+        $request = $this->application->getRequest();
```
Comment (isneep) [inline tests/unit/Validator/ManagerTest.php:85]: Mock request

File: tests/unit/Validator/ManagerTest.php [hunk 20]
```diff
@@ -50,4 +59,33 @@ public function testItHasADefinedInstanceOfAttribute(): void
...
+
+        $this->assertInstanceOf(Decode::class, $validator);
+    }
+
+    /**
+     * @return void
+     */
+    public function testItCanGetDefaultValidatorByRequest(): void
+    {
+        $this->application->setRequest('GetCurrency', ['currencyId' => 'EUR'], [], ['Currency' => []]);
```
Comment (isneep) [inline tests/unit/Validator/ManagerTest.php:83]: No application init needed

File: tests/unit/Validator/Qr/DecodeTest.php [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        $this->validator = new Decode();
+
+        $this->dummy = new DummyQr();
+    }
+
+    /**
+     * @returns void
+     */
+    private function setHydrator(): void
```
Comment (isneep) [inline tests/unit/Validator/Qr/DecodeTest.php:47]: Not necessary, already covered by the HydratorAwareTraitTest

File: tests/unit/Validator/Qr/DecodeTest.php [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+     */
+    public function testItIsAValidator(): void
+    {
+        verify($this->validator)->isInstanceOf(ValidatorInterface::class);
+    }
+
+    /**
+     * @return void
+     */
+    public function testItExtendsAbstract(): void
```
Comment (isneep) [inline tests/unit/Validator/Qr/DecodeTest.php:63]: Check it also extends RequiredMembers validator (own test method)

File: tests/unit/Validator/Qr/DecodeTest.php [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+    {
+        $this->validator->setHydrator(new ClassMethods(false, true));
+    }
+
+    /**
+     * @return void
+     */
+    public function testItIsAValidator(): void
+    {
+        verify($this->validator)->isInstanceOf(ValidatorInterface::class);
```
Comment (isneep) [inline tests/unit/Validator/Qr/DecodeTest.php:57]: Not necessary, already covered by the parent class' test

File: tests/unit/Validator/Qr/DecodeTest.php [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+        $this->setHydrator();
+
+        $this->tester->assertObjectHasMethod('isValid', $this->validator);
+        $this->tester->assertObjectMethodIsPublic('isValid', $this->validator);
+        $data = $this->tester->invokeMethod($this->validator, 'isValid', [$this->dummy]);
+        $this->assertIsBool($data);
+        $this->assertFalse($data);
+        $this->assertNotEmpty($this->validator->getMessages());
+    }
+
```
Comment (isneep) [inline tests/unit/Validator/Qr/DecodeTest.php:82]: I'm missing a test method for the validate method

File: tests/unit/Validator/Qr/EncodeTest.php [hunk 25]
```diff
@@ -0,0 +1,123 @@
+<?php
+
```
Comment (isneep) [inline tests/unit/Validator/Qr/EncodeTest.php:2]: Same remarks as for DecodeTest

**Pull request merged**